WeakAuras/Prototypes.lua

```diff
@@ -4393,159 +4393,246 @@ WeakAuras.event_prototypes = {
     events = function(trigger)
       local result = {
         "UNIT_SPELLCAST_CHANNEL_START",
-        "UNIT_SPELLCAST_CHANNEL_STOP",
-        "UNIT_SPELLCAST_CHANNEL_UPDATE",
         "UNIT_SPELLCAST_START",
-        "UNIT_SPELLCAST_STOP",
         "UNIT_SPELLCAST_DELAYED",
+        "UNIT_SPELLCAST_CHANNEL_UPDATE",
         "UNIT_SPELLCAST_INTERRUPTIBLE",
         "UNIT_SPELLCAST_NOT_INTERRUPTIBLE",
+        "UNIT_SPELLCAST_STOP",
+        "UNIT_SPELLCAST_CHANNEL_STOP",
+        "UNIT_SPELLCAST_INTERRUPTED",
+        "UNIT_SPELLCAST_FAILED",
+        "UNIT_SPELLCAST_FAILED_QUIET",
       };
-      AddUnitChangeEvents(trigger.unit, result);
-      return result;
+      AddUnitChangeEvents(trigger.unit, result)
+      if trigger.target ~= "" then
+        tinsert(result, "UNIT_TARGET")
+      end
+      return result
     end,
     internal_events = {
       "CAST_REMAINING_CHECK"
     },
     force_events = "CAST_REMAINING_CHECK",
+    canHaveAuto = true,
+    canHaveDuration = "timed",
     name = L["Cast"],
-    init = function(trigger)
-      trigger.unit = trigger.unit or "";
+    triggerFunction = function(trigger)
       local ret = [=[
-        local unit = [[%s]]
-        local inverse = %s
-        local spell, interruptible, _;
-        local castType;
-        local endTime;
-        local spellId;
-        spell, _, _, _, endTime, _, _, interruptible, spellId = UnitCastingInfo(unit)
-        if(spell) then
-          castType = "cast"
-        else
-          spell, _, _, _, endTime, _, interruptible, spellId = UnitChannelInfo(unit)
-          if(spell) then
-            castType = "channel"
+        return function(states, event, sourceUnit)
+          local trigger_inverse = %s
+          local trigger_unit = [[%s]]
+          local trigger_spellName = [[%s]]
+          local trigger_spellId = [[%s]]
+          local trigger_interruptible = %s
+          local trigger_castType = [[%s]]
+          local remainingCheck = %s
+          local trigger_target = [[%s]]
+          local trigger_clone = %s
+          local cloneId = ""
+
+          if trigger_clone and sourceUnit and UnitExists(sourceUnit) then
+            cloneId = UnitGUID(sourceUnit)
           end
-        end
-        interruptible = not interruptible;
-      ]=];
-      ret = ret:format(trigger.unit, trigger.use_inverse and "true" or "false");
 
-      if(trigger.use_remaining) then
-        local ret2 = [[
-          local expirationTime = endTime and endTime > 0 and (endTime / 1000) or 0;
-          local remaining = expirationTime - GetTime();
-          local remainingCheck = %s;
-          if(remaining >= remainingCheck and remaining > 0) then
-            WeakAuras.ScheduleCastCheck(expirationTime - remainingCheck);
+          if event == "PLAYER_TARGET_CHANGED" then
```

Comment:
I don't understand this bit, can you explain?

Reply:
"PLAYER_TARGET_CHANGED" is registered by AddUnitChangeEvents() for trigger.unit == "target" or with specific unit, for specific unit it is useless in most case except for "target-xxxx", so it might be preferable to register this event only for "target".

When this event trigger, sourceUnit is always nil and the unit that needs to be checked is trigger.unit